test/functional/command_write_concern.test.js

```diff
@@ -17,245 +17,160 @@ var extend = function(template, fields) {
-
-      // Primary server states
-      var arbiter = [
-        extend(defaultFields, {
-          ismaster: false,
-          secondary: false,
-          arbiterOnly: true,
-          me: 'localhost:32002',
-          primary: 'localhost:32000'
-        })
-      ];
-
-      // Boot the mock
-      co(function*() {
-        const primaryServer = yield mock.createServer(32000, 'localhost');
-        const firstSecondaryServer = yield mock.createServer(32001, 'localhost');
-        const arbiterServer = yield mock.createServer(32002, 'localhost');
-
-        primaryServer.setMessageHandler(request => {
-          var doc = request.document;
-
-          if (doc.ismaster) {
-            request.reply(primary[0]);
-          } else if (doc.listCollections) {
-            request.reply({
-              ok: 1,
-              cursor: {
-                id: Long.fromNumber(0),
-                ns: 'test.cmd$.listCollections',
-                firstBatch: []
-              }
-            });
-          } else if (doc.create) {
-            commandResult = doc;
-            request.reply({ ok: 1 });
-          } else if (doc.endSessions) {
-            request.reply({ ok: 1 });
-          }
-        });
-
-        firstSecondaryServer.setMessageHandler(request => {
-          var doc = request.document;
-          if (doc.ismaster) {
-            request.reply(firstSecondary[0]);
-          } else if (doc.endSessions) {
-            request.reply({ ok: 1 });
-          }
-        });
-
-        arbiterServer.setMessageHandler(request => {
-          var doc = request.document;
-          if (doc.ismaster) {
-            request.reply(arbiter[0]);
-          } else if (doc.endSessions) {
-            request.reply({ ok: 1 });
-          }
-        });
-
-        var commandResult = null;
-
-        // Connect to the mocks
-        const client = configuration.newClient(
-          'mongodb://localhost:32000,localhost:32001,localhost:32002/test?replicaSet=rs'
-        );
-
-        client.connect(function(err, client) {
-          test.equal(null, err);
-          var db = client.db(configuration.db);
+    test: writeConcernTest('aggregate', function(db, writeConcernTestOptions, done) {
+      db.collection('test')
+        .aggregate(
+          [{ $match: {} }, { $out: 'readConcernCollectionAggregate1Output' }],
+          writeConcernTestOptions
+        )
+        .toArray(done);
+    })
+  });
 
-          db.createCollection('test_collection_methods', { w: 2, wtimeout: 1000 }, function(err) {
-            test.equal(null, err);
-            test.deepEqual({ w: 2, wtimeout: 1000 }, commandResult.writeConcern);
+  it('successfully pass through writeConcern to create command', {
+    metadata: {
+      requires: {
+        generators: true,
+        topology: 'single'
+      }
+    },
 
-            client.close(done);
-          });
-        });
-      });
-    }
+    test: writeConcernTest('create', function(db, writeConcernTestOptions, done) {
```

Comment:
FWIW, I find writing helpers down to this level to be quite expedient but ultimately unreadable. I have to click through to `writeConcernTest` to figure out what's actually going on, and even there it's getting a bit "metaprogrammy" (what are we even asserting in this test right now? it's buried alive in a helper above).

I think once you have the `withStubbedClient` helper written, a little bit of copy-paste goes a long way while still removing some boilerplate. You no longer need to close a `client` when that code is in here, so the test looks very natural and readable to me:

```
withStubbedClient(function(client, commands, done) {
  const options = /* make options */;
  db.createCollection('test_collection_methods', options, err => {
    expect(err).to.not.exist;
    expect(commands).property('[0].writeConcern').to.eql({ whatever });
    done();
  });
});
```

that looks really good to me

Reply:
I'd agree if there was more variation between these tests, but they all have the exact same expectation. If the logic is repeated ten times throughout the function, I have to scroll through and verify that each test is doing the same thing when I'm reading the tests. This way, I know they are all doing the same thing, the function at the top of the file, and the precise difference between each test is extremely visible.